Add CI workflow to automate crates.io releases

funkeleinhorn opened the pull request:
This commit adds a CI workflow to automate the release of new versions to crates.io using https://lib.rs/crates/publish-action. This way commits get tagged and published automatically when a commit to the master branch changes the version string in the Cargo.toml file.

Please feel free to suggest changes :)

funkeleinhorn commented:
@zimbatm

zimbatm commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,43 @@
+          cache-name: cache-publish-action
+        with:
+          path: ~/.cargo
+          key: ${{ runner.os }}-build-${{ env.cache-name }}-v0.2.0
```
what is the v0.2.0 for?

funkeleinhorn replied:
Removed it as it seems unnecessary. My guess is to have a cache per version of the cargo publish action.

zimbatm commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,43 @@
+      - name: Install publish-action
+        if: steps.cache-publish-action.outputs.cache-hit != 'true'
+        run:
+          cargo install publish-action --version=0.2.0
```
Is the version hard-coded here?

funkeleinhorn replied:
Removed the version hard code to use the latest version instead.

zimbatm commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,43 @@
+
+on:
+  push:
+    branches: [ master ]
```
It might be best to only publish on releases instead of on every push to master

funkeleinhorn replied:
Currently there are no releases just tagged commits. The CI is currently configured to auto tag a commit to master where the version in the Cargo.toml changes and publishes it to crates.io. This way we can not forget to tag releases. But I can understand if you prefer the workflow of manually tagging commits and the tagging of a commit triggers it to be published. If you want I can change the CI to do no tagging on the repo and just publish new version tags to crates.io

funkeleinhorn commented:
I changed the CI script to be triggered by new tags matching a version instead of every new commit. This way the CI also does not need write access to the repo which seems nice security wise. I also replaced the cargo-publish-action crate with using cargo publish directly.

zimbatm commented:
ok I added the secret so that *should* work. I might ping you again if it breaks at the next release :)
